### ISSUE TITLE
[🐛 BUG] Error message when creating Scenarios with scenario_selector

### ISSUE THREAD

**Issue by enarroied** [Avaiga/taipy#2009]
### What went wrong? 🤔

I just installed Taipy 4.0.
When I use the scenario_selector and I create a new Scenario, I get this on the terminal:
```
C:\Users\PC\anaconda3\envs\taipy\Lib\site-packages\taipy\gui\gui.py:821: TaipyGuiWarning: :
Traceback (most recent call last):
  File "C:\Users\PC\anaconda3\envs\taipy\Lib\site-packages\taipy\gui\gui.py", line 819, in _call_on_change
    var_name, current_context = self._get_real_var_name(var_name)
                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\PC\anaconda3\envs\taipy\Lib\site-packages\taipy\gui\gui.py", line 806, in _get_real_var_name
    raise NameError(f"Can't find matching variable for {var_name} on context: {current_context}")
NameError: Can't find matching variable for selected_scenario on context: pages.scenario_selector

  _warn("", e)
```

**However, the scenario is created.**

### Expected Behavior

If there are no errors... then it shouldn't display an error message. I haven't seen any malfunction (but maybe there is one, somewhere...)

### Steps to Reproduce Issue

The scenario_selector looks like this:
```
    tgb.scenario_selector(
        value="{selected_scenario}",
        show_pins=True,
        on_change=notify_change,
        on_creation=notify_creation,
    )
```

### Runtime Environment

Windows 10, I used Python 12 within Anaconda, Taipy develop

### Browsers

Brave

### OS

Windows

### Version of Taipy

4.0.0

### Acceptance Criteria

- [ ] Ensure new code is unit tested, and check code coverage is at least 90%.
- [ ] Create related issue in taipy-doc for documentation and Release Notes.

### Code of Conduct

- [X] I have checked the [existing issues](https://github.com/Avaiga/taipy/issues?q=is%3Aissue+).
- [ ] I am willing to work on this issue (optional)

**Comment by akshatshukla13** [comment]
Would like to work on this issue. Assign it to me!

**Comment by FredLL-Avaiga** [comment]
I tried with
```py
import taipy.gui.builder as tgb
from taipy import Config
from taipy.gui import Gui, State


def notify_change(state: State, var: str, val: str):
    print(f"notify_change(state: State, var: {var}, val: {val})")


def notify_creation(state: State, var: str, val: str):
    print(f"notify_creation(state: State, var: {var}, val: {val})")


selected_scenario = None

# Configuration of scenario
scenario_cfg = Config.configure_scenario(id="my_scenario")

with tgb.Page() as page:
    tgb.scenario_selector(
        value="{selected_scenario}",
        show_pins=True,
        on_change=notify_change,
        on_creation=notify_creation,
    )

if __name__ == "__main__":
    Gui(page).run(title="2009 [🐛 BUG] Error message when creating Scenarios with scenario_selector")

```

and it works for me (tm)

```
[2024-10-14 18:12:17.455][Taipy][INFO] 'allow_unsafe_werkzeug' has been set to True
[2024-10-14 18:12:17.455][Taipy][INFO] 'async_mode' parameter has been overridden to 'threading'. Using Flask built-in development server with debug mode
[2024-10-14 18:12:20.090][Taipy][INFO]  * Server starting on http://127.0.0.1:5000
[2024-10-14 18:12:20.091][Taipy][INFO]  * Application is accessible at http://localhost:5000
 * Serving Flask app 'Taipy'
 * Debug mode: on
notify_creation(state: State, var: , val: {'action': 'notify_creation', 'config': <taipy.core.config.scenario_config.ScenarioConfig object at 0x00000147C076C910>, 'date': datetime.datetime(2024, 10, 14, 16, 12, 23, 805000, tzinfo=tzutc()), 'label': 'qqq', 'properties': {}})
[2024-10-14 18:12:42.810][Taipy][INFO] Updating configuration with command-line arguments...
[2024-10-14 18:12:42.811][Taipy][INFO] Managing application's version...
[2024-10-14 18:12:42.813][Taipy][INFO] Development mode: Clean all entities of version ac105463-6ca0-4c54-abc5-fb9c3ad4b101
[2024-10-14 18:12:42.948][Taipy][INFO] Checking application's version...
[2024-10-14 18:12:42.951][Taipy][WARNING] tasks field of ScenarioConfig `my_scenario` is empty.
[2024-10-14 18:12:42.952][Taipy][INFO] Blocking configuration update...
notify_change(state: State, var: selected_scenario, val: qqq)

```
![image](https://github.com/user-attachments/assets/825af051-f2dc-4730-9e5c-643a2dd950e8)

**Comment by FabienLelaquais** [comment]
@FlorianJacta @AlexandreSajus Can you confirm this is a problem, or not?

**Comment by FlorianJacta** [comment]
I confirm this is an issue. @FabienLelaquais @FredLL-Avaiga I'll try to reproduce it for you with a little code

**Comment by jrobinAV** [comment]
We need a piece of code that reproduces the bug. Otherwise, we cannot investigate. @enarroied @FlorianJacta can one of you provide some code 😉? 

Thanks

**Comment by enarroied** [comment]
Hi everyone,
So I think I have something interesting. The problem seems to happen when the Scenario is imported to a file that is then called by main.py.

Example:
* File 1: `orchestration.py`:

```python
import taipy as tp
from taipy import Config, Orchestrator

node_config = Config.configure_data_node(id="test_node")

scenario_config = Config.configure_scenario(
    id="test_scenario",
    additional_data_node_configs=node_config,
)

orchestrator = Orchestrator()
orchestrator.run()

test_scenario = tp.create_scenario(scenario_config)
```

* File 2: `pages/scenario.py`:

```python
import taipy.gui.builder as tgb
from orchestration import test_scenario

selected_scenario = test_scenario


def notify_change(state, var, val):
    print(f"notify_change(var: {var}, val: {val})")


def notify_creation(state, var, val):
    print(f"notify_creation(var: {var}, val: {val})")


with tgb.Page() as buggy_page:
    tgb.scenario_selector(
        value="{selected_scenario}",
        show_pins=True,
        on_change=notify_change,
        on_creation=notify_creation,
    )
```
* File 3: `main.py`:
```python
from pages.scenario import buggy_page
from taipy.gui import Gui

if __name__ == "__main__":
    Gui(page=buggy_page).run(
        title="2009 [🐛 BUG] Error message when creating Scenarios with scenario_selector",
        use_reloader=True,
    )
```

------

If you take the following to `main.py`, there is no more error:
```python
from orchestration import test_scenario

selected_scenario = test_scenario`
```

------

So I don't know if it's a real bug, or if I need to understand how to make the Orchestrator interact with the front end.

I hope this helps, it was a nice investigation 😄

**Comment by FredLL-Avaiga** [comment]
Tx @enarroied, I can replicate the bug now

**Comment by jrobinAV** [comment]
@enarroied,
Hi Eric, 

I have a remark that is out of scope and on another topic.

We generally recommend running the Taipy services in the main script under the `if __name__ == "__main__":` block. Something like the following:

* File 1: `config.py`:

```python
import taipy as tp
from taipy import Config

node_config = Config.configure_data_node(id="test_node")

scenario_config = Config.configure_scenario(
    id="test_scenario",
    additional_data_node_configs=node_config,
)
```
* File 2: `pages/scenario.py`:

```python
import taipy.gui.builder as tgb

selected_scenario = None


def notify_change(state, var, val):
    print(f"notify_change(var: {var}, val: {val})")


def notify_creation(state, var, val):
    print(f"notify_creation(var: {var}, val: {val})")


with tgb.Page() as buggy_page:
    tgb.scenario_selector(
        value="{selected_scenario}",
        show_pins=True,
        on_change=notify_change,
        on_creation=notify_creation,
    )
```
* File 3: `main.py`
```python
from pages.scenario import buggy_page, selected_scenario
from config import scenario_config
from taipy.gui import Gui

if __name__ == "__main__":
    Orchestrator().run() 
    selected_scenario = tp.create_scenario(scenario_config)

    Gui(page=buggy_page).run(
        title="2009 [🐛 BUG] Error message when creating Scenarios with scenario_selector",
        use_reloader=True,
    )
```

The reason is that the main.py script is executed every time a sub-process is spawned by your Python program. It includes your Python code, the Taipy code, or any Python library you use. The code under the `if __name__ == "__main__"` block will **NOT** be re-executed, though.

If you don't run the service and create your scenario within the if block, it will be executed at the start of each sub-process. This results in starting the orchestrator multiple times and creating a scenario multiple times. 

In Taipy, some sub-processes are spawned if you run in standalone mode (to execute the tasks in parallel).

**Comment by enarroied** [comment]
Hi @jrobinAV 
Thank you very much for this explanation. 
I'll make some changes in how I structure my apps, this will be very useful, I'm sure!

**Comment by AlexandreSajus** [comment]
Fixed in 4.0.1.dev1
🎉
